UI: Add a verbose debug logging checkbox next to "Report a bug"

NiLuJe opened the pull request:
@hius07 mentioned something to that effect a few weeks back, I liked it ;p.

To make it slightly more user-friendly (and warrant the code duplication ;p), this always flips debug + verbose at once, and looks disabled if you only have debug w/o verbose enabled. It also requests a restart, for framebuffer's sake (it pulls a ref to logger.dbg on init, but Logger mutates its methods at runtime depending on the log level, so this doesn't track updates to the logging level).

(Also implement an actual `dbg.log` function that *will* keep track of the current status, in case something like that comes up again in the future).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/11366)
<!-- Reviewable:end -->

Frenzie commented:
Screenshot:
![image](https://github.com/koreader/koreader/assets/202757/e122a41a-9189-4cca-b850-48a280215845)

It's… a little ugly tbh. What if you make "Report a bug" a confirmbox to flip it?

NiLuJe commented:
> Would you guys like to change the Bug report template in the same PR as well?
> 
> So that it'll point to where it will be agreed (just a reminder)

Can't do that (backward compat, it's not actually there yet ;p). It'll have to wait until this hits a couple of stables at least.

poire-z commented on frontend/ui/elements/common_info_menu_table.lua:
```diff
@@ -51,38 +52,17 @@ common_info.about = {
         local common_msg = T(_("Please report bugs to \nhttps://github.com/koreader/koreader/issues\n\nVersion:\n%1\n\nDetected device:\n%2"),
             Version:getCurrentRevision(), Device:info())
-        local log_msg = T(_("Reproduce the issue with verbose debug logging enabled, and attach %1 to your bug report."), log_path)
+        local log_msg = T(_("Closing this popup will toggle verbose debug logging, reproduce the issue while it's enabled, and attach %1 to your bug report."), log_path)
```
Feels a bit authoritarian to force that on a user :) I could just tap tihs menu item to see what's proposed - and I'll get stuck with verbose debug logs, that I'll have to find out how to disable (or not care and fill my device with logs) :/

(And somehow, not all report needs a crash.log - and we'll get 5Mb crash.log that we will have to look at even if you guess from the issue description they won't give any insight, just in case.)

NiLuJe replied:
> Regardless of that, yeah, I'm not entirely sold on merging those two myself, but I get @Frenzie's original concern, too; and I don't think adding yet another prompt to make it optional would be any better...

I just noticed that I misread Frenzie's poposition though, which was a confirmbox, that I somehow read as a checkmark, for some reason xD.

That probably makes more sense, going to try that ;).

NiLuJe replied:
Now looks like this: 
![unibutton](https://github.com/koreader/koreader/assets/111974/51e7960e-9a60-420e-b37a-52a4ae5c0c9e)

I went with an unibutton because it looked cramped otherwise, and we still have ample space to tap outside to dismiss.

NiLuJe replied:
Not sure if it wouldn't be clearer to have a `text_func` to append a spelled out "(logging enabled)" instead of the checkmark?

NiLuJe replied:
> Not sure if it wouldn't be clearer to have a `text_func` to append a spelled out "(logging enabled)" instead of the checkmark?

Eeh, not a fan.

```diff
diff --git a/frontend/ui/elements/common_info_menu_table.lua b/frontend/ui/elements/common_info_menu_table.lua
index 6d85941fb..b4b2562c8 100644
--- a/frontend/ui/elements/common_info_menu_table.lua
+++ b/frontend/ui/elements/common_info_menu_table.lua
@@ -54,10 +54,19 @@ common_info.about = {
     end
 }
 common_info.report_bug = {
-    text = _("Report a bug"),
+    text_func = function()
+        local label = _("Report a bug")
+        if G_reader_settings:isTrue("debug_verbose") then
+            label = label .. " " .. _("(verbose logging enabled)")
+        end
+        return label
+    end,
+    keep_menu_open = true,
+    --[[
     checked_func = function()
         return G_reader_settings:isTrue("debug_verbose")
     end,
+    --]]
     callback = function(touchmenu_instance)
         local DataStorage = require("datastorage")
         local log_path = string.format("%s/%s", DataStorage:getDataDir(), "crash.log")
```

Frenzie commented on frontend/dbg.lua:
```diff
@@ -86,6 +86,13 @@ function Dbg:v(...)
 end
 
+--- Conditional logging with a stable ref.
+function Dbg.log(...)
+    if Dbg.is_on then
+        return LvDEBUG(...)
+    end
+end
```
Would it make sense to switch references around?

Dbg.log = function() end

and

Dbg.log = Dbg.forRealLvDEBUG

To avoid the extra condition.

NiLuJe replied:
The condition is kind of the point, though.

The intended use-case is for funky stuff like what framebuffer does, where the class has a `debug` function that is NOP, that is replaced during instantiation by a reference to `logger.dbg` fed by the constructor.

The issue is that Logger mutates its methods, so if debugging is disabled, that's NOP, and that stale reference will stay pinned to NOP forever ;).

If you feed it that instead, it works as intended: logging follows the actual debugging state.

If it was instead declared as `Dbg.log = LvDEBUG` (or wrapped in a useless closure as `function Dbg.log = function(...) return LvDEBUG(...) end`, it would log *always*, even if debug is disabled.

(For context, `local LvDEBUG = logger.LvDEBUG`, and that's a Logger method that is *never* mutated, and always logs in under the "dbg" context).

NiLuJe replied:
(Unless I completely missed your point, that how I understood the above ;)).

Frenzie replied:
My point was why only mutate on instantiation if this is desired functionality?

Frenzie replied:
Dbg:turnOn() and turnOff() already do this for guard() and dassert(). Is it so bad to mutate if the alternative is to always branch?

NiLuJe replied:
But `guard` & `dassert` are always called via `dbg.`, we never store a reference to them in an assignment and then use *that* stale reference to call them, which is the whole difference. So we effectively always look up the current, live ref.

To be clear: I *like* the current behavior, despite its mild inconvenience of requiring a restart because of framebuffer's pinned ref.

I just wanted to provide an alternative if we ever needed something like that in the future, at the cost of a closure & a branch ;).

NiLuJe replied:
Which leads me to a possibly much simpler solution: pass a ref to logger, instead of logger.dbg, and then just call logger.dbg(), that'll look up the current ref.

NiLuJe replied:
> Which leads me to a possibly much simpler solution: pass a ref to logger, instead of logger.dbg, and then just call self.logger.dbg(), that'll look up the current ref.

Eh, that would probably make the testsuite cranky, unless we implement a stub logger class with all the right methods :/.

Frenzie replied:
> To be clear: I _like_ the current behavior, despite its mild inconvenience of requiring a restart because of framebuffer's pinned ref.

I'm not the one complaining about a restart; you're the one adding this branch here, lol

poire-z commented on frontend/apps/reader/modules/readerfooter.lua:
```diff
@@ -591,7 +591,7 @@ function ReaderFooter:init()
 
     self.text_container = RightContainer:new{
-        dimen = Geom:new{ w = 0, h = self.height },
+        dimen = Geom:new{ x = 0, y = 0, w = 0, h = self.height },
```
It feels like I've seen that before - either you already did it, or you tried to and we may have discussed it :)

Anyway, this still feel strange to me, having to put these x/y where they are not needed and make not the sense they'd have if explicitely written - because we're not putting that stuff at x=0 y=0. And I'll probably forget to write them :)
If they are needed (ie. to make 4 slots in the table object and avoid Lua having to resize that table), it feels it should be internal to `Geom:new/init()`.

NiLuJe replied:
> Geom:new/init()

That would be too late. The point is to make the initial table the "right" size, and that's the table being fed to new ;).

Note that some of these are quite likely overly pedantic, I just blindly filled 'em up whenever I encountered one.

poire-z replied:
pedantic/blindly vs. clarity/expliciteness: no x/y has always meant it's a not-(yet-)positionned dimen.
Is this really just to avoid a Lua table resize ? Is that really necessary vs. the unexpiciteness? Are you sure a table resize, and how costly, happens when adding 2 items to a initial 2-items table?

NiLuJe replied:
> no x/y has always meant it's a not-(yet-)positionned dimen

To me it often looks like "uh, I dunno how this really works, so I'm going to leave that alone ;p" ^^.

If you find that (0, 0) is confusing, we could use a sentinel value instead (I dunno, -1 or 0xFFFF or something). I would tend to prefer (0, 0), though (if only because it feels like zero-initializing in C/C++).

> Are you sure a table resize, and how costly, happens when adding 2 items to a initial 2-items table?

Well, it's a `realloc` that's easily avoidable. I don't get the readability concerns myself (as I mentioned above, I find *notù specifying the coords *more* confusing and wafflish), but they should trump everything else, so, RFC.

hius07 replied:
Can it be instead of
https://github.com/koreader/koreader/blob/e0ed04a1a147162c0c03e5864df2e62e7a5b9ba2/frontend/ui/geometry.lua#L30-L42
```Lua
local Geom = {}

function Geom:new(o)
    local new_o = {
        x = o.x or 0, -- left origin
        y = o.y or 0, -- top origin
        w = o.w or 0, -- width
        h = o.h or 0, -- height
    }
    setmetatable(new_o, self)
    self.__index = self
    return new_o
end

```
And do not pass default zeroes. Initial table in the caller will be gc'ed. No table resizing.

NiLuJe replied:
Interesting thought, but that would cause GC churn instead, which I'm not convinced would be a great trade ;).

On the other hand, for completely empty Geoms,

```diff
diff --git a/frontend/ui/geometry.lua b/frontend/ui/geometry.lua
index 7c5d27197..792806bca 100644
--- a/frontend/ui/geometry.lua
+++ b/frontend/ui/geometry.lua
@@ -35,7 +35,12 @@ local Geom = {
 }
 
 function Geom:new(o)
-    if not o then o = {} end
+    if not o then
+        o = {
+            x = 0, y = 0,
+            w = 0, h = 0,
+        }
+    end
     setmetatable(o, self)
     self.__index = self
     return o
```

makes sense. Doesn't really address (most of) what we're talking, about though ;p.

hius07 replied:
The purpose is to avoid passing `x = 0, y = 0`.

NiLuJe replied:
> The purpose is to avoid passing `x = 0, y = 0`.

I get that, but it ends up making *everything* worse from a performance standpoint, compared to only *sometimes* maybe needing a realloc or not depending on the constructor table ;).

poire-z replied:
No other good idea :/
(A bad one, requiring a function call instead: `function Geom.getNewGeom(w, h, x, y) ...` to which we could not provide x/y and that would build and return one table here.)

> making everything worse from a performance standpoint

How much would that be vs. all the other things happening when Geom objects are created, = when widgets are created: there's probably a lot of other objects created/resized, + repaints/filling blitbuffer/refreshs so that the realloc is negligible :)

(But there's not much of my widgets that get their Geom stuff fixed in the current state of this PR :) so may be I did that right (I don't think so :)) or your eyes and mind were not that strong, and you will allow some uncomplete ones to go by if not in hot codepaths - and I would compromise on letting you add them to bits you consider hot-widget-paths, but not making that a hard rule :))

NiLuJe replied:
> How much would that be vs. all the other things happening

Probably noise. But my point was more that it would somewhat gratuitously make *every* caller "worse" vs. the current status quo or only *some* (and even for those, that's a mild maybe: there's definitely going to be inheritance lookups, but actual writes? That's somewhat rarer).

NiLuJe replied:
There, left classes that don't actually care about dimensions alone, massively reducing the amount of changes ;).

NiLuJe reviewed (commented):
> Reviewed 3 of 6 files at r1, 3 of 19 files at r4, 2 of 16 files at r5, 16 of 16 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11366)* status: all files reviewed, 3 unresolved discussions (waiting on @Frenzie, @pazos, and @poire-z)



<!-- Sent from Reviewable.io -->

poire-z commented on frontend/ui/elements/common_info_menu_table.lua:
```diff
@@ -95,8 +76,31 @@ common_info.report_bug = {
         else
             msg = common_msg
         end
-        UIManager:show(InfoMessage:new{
+        UIManager:show(ConfirmBox:new{
             text = msg,
+            no_cancel_button = true,
+            ok_text = _("Toggle verbose logging"),
```
Well, I won't see that ConfirmBox often, and so won't most people. So, is does not need to be pretty nor ugly - but it would be nice if it still has a `Cancel` button like most/all others, just so people not familiar with dissmis by tap outside don't get there wondering how to get out, and taping the only thing they think they can tap: Toggle verbose logging :)

ConfirmBox can have 2 wide buttons (in case you removed Cancel because it was taking room needed by the wide "Toggle verbose logging":
https://github.com/koreader/koreader/blob/e0ed04a1a147162c0c03e5864df2e62e7a5b9ba2/frontend/ui/widget/confirmbox.lua#L52-L53

hius07 replied:
Is question icon needed? (no questions in the text)
Maybe, ButtonDialog? Without changes to the ConfirmBox widget.

Frenzie replied:
The cancel button for dismissal was included in my suggestion, but the fact that ConfirmBox comes with an icon is something I wasn't thinking of. :-)

Btw, I'm not sure how much sense a checkmark makes there in the menu?

NiLuJe replied:
> ConfirmBox can have 2 wide buttons (in case you removed Cancel because it was taking room needed by the wide "Toggle verbose logging":

Oh, I looked at that feature *before* coming up with removing the Cancel, and then forget about it ;p. Two rows of wide buttons, with cancel at the bottom would probably be okay, yeah.

> Is question icon needed?

That's a fair point. I honestly don't even notice the icons anymore in those popups ;p.

> Maybe, ButtonDialog? 

I thought about that, but thought I'd have to reinvent the wheel a bit too much for that one.

> Btw, I'm not sure how much sense a checkmark makes there in the menu?

The idea is to have *something* to show that verbose debug logging is enabled at a glance.
I'm not super happy about it, but so far I find that it works better compared to the only other solution I could think of, which was a [`text_func`](https://github.com/koreader/koreader/pull/11366#discussion_r1454340103).

NiLuJe replied:
> > Maybe, ButtonDialog?
> 
> I thought about that, but it looked like I'd have to reinvent the wheel a bit too much for that one.

Specifically, that I'd have to stick some kind of text widget to it to actually display the most important part of the original and main intent of it ;).

poire-z commented on frontend/ui/widget/container/widgetcontainer.lua:
```diff
@@ -55,6 +55,7 @@ function WidgetContainer:paintTo(bb, x, y)
         self.dimen = Geom:new{x = 0, y = 0, w = content_size.w, h = content_size.h}
     end
 
+    -- NOTE: Clunky or left in on the off-chance we're passed a dimen that isn't a proper Geom object...
```
(Spent 30 seconds reading this :) I guess (right?) a comma between `in` and `on` would have reduced it to 3s :)

NiLuJe replied:
Yeah, I should have backticked the `or` to make it clearer ;).

poire-z commented:
> There, left classes that don't actually care about coordinates alone, massively reducing the amount of changes ;).

Thanks, that's bearable :)
(I accept the one in TextBoxWidget because we have so many of them.)

For the other stuff, some questions:
You added x=0, y=0 - and dimen=dimen:copy() to most of the stuff related to the 2 reported issues: keyboard shortcut letter square badly positionned, and menu items not tapable on the full width.
Do you think there will be others that we'll meet and that will need that treatment, or have you done a visit (or the right grep'ing) of all the stuff and it feels it's all ok?
And if grep, what pattern ? And even if not: any rule of thumb for us later to guess that we need to specify x/y=0 (I guess there are not really needed, just your performance reasons?) and when to use :copy() (seems related to a prototype-object we create and that we clone?)

NiLuJe commented:
> Do you think there will be others that we'll meet and that will need that treatment, or have you done a visit (or the right grep'ing) of all the stuff and it feels it's all ok?

Yup, they all mostly feed a `Geom:new{...}` to it, these were the only ones passing one by reference (there may be one or two callers that do pass a ref, but it isn't shared with anything).

> And if grep, what pattern ?

`rg -A15 'FrameContainer:new' | rg 'dimen[[:blank:]]*='`

(Same idea for the other widgets that were fixed, but they had far less call sites, and none of them even remotely problermatic).

> And even if not: any rule of thumb for us later to guess that we need to specify x/y=0

I double-checked the `paintTo` method of the subclass to see if they did anything with coordinates (some don't even use `self.dimen` at all). (Also keeping in mind some special snowflakes like `OverlapGroup`, that will *use* part of the constructor's dimen, but create a new one based on it at *init*, and never touch it again).

As an actual recommendation with some semantic value to it, I'd say if you're dealing with an actual rectangle, setup coordinates at construct, but if you're not (e.g., LineWidget or *Spans), you're probably fine ;).

> when to use :copy()

That one's easy: assume *every* widget will use `self.dimen` by reference and will keep that reference alive, so, don't share it with any other widget unless you know how they might interact with each other ;).

(So, whether you copy or not essentially depends how you're building your widget and that geom object; the take-away is that each widget should have its own geom object).

NiLuJe commented:
> menu items not tapable on the full width

Oh, that one was actually because of `paintTo` methods overwriting `dimen.w` & `dimen.h` (which is something I added as a separate commit during review because I thought it might help *avoid* potential wonky call-sites ;p. Wrong ^^).

(It was also breaking another semi-reliable contract with dimen & our widgets, where if actually set the dimensions, it means you want those dimensions to stick, whatever the actual size of the content, so it mades perfect sense to drop it, that's how the rest of our sane widgets behave).

NiLuJe commented:
Addressed most review comments, the only question left is about the checkmark vs. text_func (or whatever else you cna think of to indicate that logging is enabled) ;).

Now looks like this:

![doublewide](https://github.com/koreader/koreader/assets/111974/53943e82-e902-4540-a98d-ab5e193c9698)

poire-z commented:
> the only question left is about the checkmark vs. text_func (or whatever else you can think of to indicate that logging is enabled) ;).

Seeing the checkbox (and half the text, masked by the dialog), it feels it could be mistaken for some kind of automatic reporting of bugs, something scary that will record what I do and read and send it to developpers as evil as Microsoft who did that first :)
So, yes, a text_func to append `[✔ verbuzz logging]` to the normal text would feel better.

NiLuJe commented:
> text_func

Done!

NiLuJe reviewed (commented):
> Reviewed 18 of 20 files at r8, 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11366)* status: all files reviewed, 5 unresolved discussions (waiting on @Frenzie, @pazos, and @poire-z)



<!-- Sent from Reviewable.io -->

Frenzie commented:
Cheers, saved me the trouble.

poire-z commented on frontend/ui/elements/common_info_menu_table.lua:
```diff
@@ -71,8 +80,37 @@ common_info.report_bug = {
+                            dbg:setVerbose(true)
+                            G_reader_settings:makeTrue("debug")
+                            G_reader_settings:makeTrue("debug_verbose")
+                            Notification:notify(_("Verbose logging enabled"), Notification.SOURCE_ALWAYS_SHOW)
```
(Not really caring, but do you need these notifications now that you have it written in the menu item just under this?)

poire-z reviewed (approved):
If you plan to rebase and keep multiple commits, may be squash the 2nd and 3rd that reverts much of the 2nd.

NiLuJe commented:
> If you plan to rebase and keep multiple commits, may be squash the 2nd and 3rd that reverts much of the 2nd.

I'd kinda like to keep it for archeological purposes ;p.
